server/mqtt.go

```diff
@@ -3835,6 +4299,102 @@ func (sess *mqttSession) cleanupFailedSub(c *client, sub *subscription, cc *Cons
 	}
 }
 
+// Make sure we are set up to deliver PUBREL messages to this QOS2-subscribed
+// session.
+//
+// Session lock held on entry. Need to make sure no other subscribe packet races
+// to do the same.
+func (sess *mqttSession) ensurePubRelConsumerSubscription(c *client) error {
+	deliverSubj := sess.pubRelDeliverySubject()
+	deliverSubjB := []byte(deliverSubj)
+	opts := c.srv.getOpts()
+	ackWait := opts.MQTT.AckWait
+	if ackWait == 0 {
+		ackWait = mqttDefaultAckWait
+	}
+	maxAckPending := int(opts.MQTT.MaxAckPending)
+	if maxAckPending == 0 {
+		maxAckPending = mqttDefaultMaxAckPending
+	}
+
+	// Subscribe before the consumer is created so we don't loose any messages.
+	if !sess.pubRelSubscribed {
+		_, err := c.processSub(deliverSubjB, nil, deliverSubjB, mqttDeliverPubRelCb, false)
+		if err != nil {
+			c.Errorf("Unable to create subscription for JetStream consumer on %q: %v", deliverSubj, err)
+			return err
+		}
+		sess.pubRelSubscribed = true
+	}
+
+	// Create the consumer if needed.
+	if sess.pubRelConsumer == nil {
+		// Check that the limit of subs' maxAckPending are not going over the limit
+		if after := sess.tmaxack + maxAckPending; after > mqttMaxAckTotalLimit {
+			return fmt.Errorf("max_ack_pending for all consumers would be %v which exceeds the limit of %v",
+				after, mqttMaxAckTotalLimit)
+		}
+
+		ccr := &CreateConsumerRequest{
+			Stream: mqttQOS2PubRelStreamName,
+			Config: ConsumerConfig{
+				DeliverSubject: deliverSubj,
+				Durable:        sess.idHash + "_pubrel",
+				AckPolicy:      AckExplicit,
+				DeliverPolicy:  DeliverNew,
+				FilterSubject:  sess.pubRelSubject(),
+				AckWait:        ackWait,
+				MaxAckPending:  maxAckPending,
+				MemoryStorage:  opts.MQTT.ConsumerMemoryStorage,
+			},
+		}
+		if opts.MQTT.ConsumerInactiveThreshold > 0 {
+			ccr.Config.InactiveThreshold = opts.MQTT.ConsumerInactiveThreshold
+		}
+		if _, err := sess.jsa.createConsumer(ccr); err != nil {
+			c.Errorf("Unable to add JetStream consumer for PUBREL for client %q: err=%v", sess.id, err)
+			return err
+		}
+		sess.pubRelConsumer = &ccr.Config
+		sess.tmaxack += maxAckPending
+	}
+
+	return nil
+}
+
+// func (sess *mqttSession) countQOS2Subs() int {
```

Comment:
So you are going to invoke that in another commit? If not then remove the code (copy somewhere if needed)

Reply:
Still uncertain if there's ever a use-case for removing the PubRel consumer once there are no QoS2 subs. Probably not since there may be pending PUBRELs there, will remove.